tools/cache_creator/cache_creator.cpp

```diff
@@ -23,12 +23,165 @@
+// @returns : UUID string in the printable format
+UuidString uuidToHexString(llvm::ArrayRef<uint8_t> uuid) {
+  assert(uuid.size() == VK_UUID_SIZE);
+
+  UuidString res;
+  llvm::raw_svector_ostream os(res);
+  for (auto groupSize : {4, 2, 2, 2, 6}) {
+    os << llvm::format_bytes(uuid.take_front(groupSize), llvm::None,
+                             /* NumPerLine = */ 16, /* BytesPerLine = */ 6);
+    uuid = uuid.drop_front(groupSize);
+    if (!uuid.empty())
+      os << '-';
+  }
+  assert(isValidHexUUIDStr(res));
+  return res;
+}
+
+static uint8_t hexDigitToNum(char c) {
+  assert((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f'));
+  return (c >= '0' && c <= '9') ? c - '0' : 10 + (c - 'a');
+}
+
+// =====================================================================================================================
+// Converts a UUID string into its binary representation.
+//
+// @param hexStr : The input hex string in the UUID format
+// @param [out] outUuid : The output buffer where de-serialized UUID gets written to
+// @returns : `true` on success, `false` when the passed `hexStr` is not valid
+bool hexStringToUuid(llvm::StringRef hexStr, llvm::MutableArrayRef<uint8_t> outUuid) {
+  assert(outUuid.size() == VK_UUID_SIZE);
+  if (!isValidHexUUIDStr(hexStr))
+    return false;
+
+  // Go over all hex digits and write out a nibble at a time.
+  // Updates `outUuid` so that the current byte is always at the front.
+  for (auto &&posAndHexDigit : llvm::enumerate(llvm::make_filter_range(hexStr, [](char c) { return c != '-'; }))) {
+    const uint8_t num = hexDigitToNum(posAndHexDigit.value());
+    if (posAndHexDigit.index() % 2 == 0) {
+      outUuid.front() = num << 4;
+    } else {
+      outUuid.front() |= num;
+      outUuid = outUuid.drop_front();
+    }
+  }
+
+  return true;
+}
+
+// =====================================================================================================================
+// Extracts cache hash from the given elf file. Looks for the llpc_cache_hash note in note sections.
+//
+// @param elfBuffer : The input elf file
+// @returns : The hash found on success, or error on failure
+llvm::Expected<Util::MetroHash::Hash> getElfCacheHash(llvm::MemoryBufferRef elfBuffer) {
+  auto elfObjectOrErr = llvm::object::ELF64LEObjectFile::create(elfBuffer);
+  if (auto err = elfObjectOrErr.takeError())
+    return std::move(err);
+
+  using ElfFileTy = llvm::object::ELF64LEFile;
+  const ElfFileTy &elfFile = elfObjectOrErr->getELFFile();
+
+  auto sectionsOrErr = elfFile.sections();
+  if (auto err = sectionsOrErr.takeError())
+    return std::move(err);
+
+  for (const auto &sectionHeader : *sectionsOrErr) {
+    if (sectionHeader.sh_type != llvm::ELF::SHT_NOTE)
+      continue;
+
+    auto sectionNameOrErr = elfFile.getSectionName(sectionHeader);
+    if (auto err = sectionNameOrErr.takeError()) {
+      llvm::consumeError(std::move(err));
+      continue;
+    }
+    if (!sectionNameOrErr->startswith(".note"))
```

Comment:
Ack. I think this check should be fine for now as it only checks for the prefix instead of an exact match. I expect the final name to start with `.note`, and we can update this to the exact name once the LLPC stuff gets merged.

Reply:
Sure, this will work for now.  It may just look at more sections than necessary.